prov/shm/src/smr_init.c

```diff
@@ -168,6 +168,9 @@ static int smr_getinfo(uint32_t version, const char *node, const char *service,
 		if (cur->caps & FI_HMEM) {
 			if (!(mr_mode & FI_MR_HMEM)) {
 				fi_freeinfo(cur);
+				if (!(hints->caps & FI_HMEM))
+					continue;
+
```

Comment:
I still don't think this is correct. The problem is that 1 of shm's fi_info structures may match and still need to be returned.  But this path returns a failure.  And it frees the whatever cur is, plus anything beyond cur, but if there's an fi_info before cur, that will still point to this freed structure.

As @j-xiong mentioned in the issue, you would need to pluck this entry out of the list, but still return the ones that are valid.  Or fail ever adding it to the list in the first place, which is easier.

Reply:
Agree. This patch would cause more trouble than it fixes. Letting util_getinfo() to filter out the unwanted entries is simpler.

Reply:
See #7777 for an alternate solution